framework/finalizer.go

```diff
@@ -0,0 +1,99 @@
+package framework
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/rest"
+)
+
+const (
+	finalizerName = "operatorkit.giantswarm.io"
+)
+
+type patchSpec struct {
+	Op    string   `json:"op"`
+	Path  string   `json:"path"`
+	Value []string `json:"value"`
+}
+
+func (f *Framework) addFinalizer(obj interface{}) (bool, error) {
+	accessor, err := meta.Accessor(obj)
+	if err != nil {
+		return false, err
+	}
+	if accessor.GetDeletionTimestamp() != nil {
+		return true, nil // object has been marked for deletion, we should ignore it.
+	}
+	if containsFinalizer(accessor.GetFinalizers(), finalizerName) {
+		return false, nil // object already has the finalizer.
+	}
+
+	c, err := rest.RESTClientFor(&rest.Config{})
+	if err != nil {
+		return false, err
+	}
+	patch := fmt.Sprintf(`{"op":"add","value":%q,"path":"/metadata/finalizers/1"}`, finalizerName)
+	res := c.Patch(types.JSONPatchType).
+		AbsPath(accessor.GetSelfLink()).
+		Body(patch).
+		Do()
+	if res.Error() != nil {
+		return false, res.Error()
+	}
+
+	return true, nil
+}
+
+func (f *Framework) removeFinalizer(obj interface{}) error {
+	accessor, err := meta.Accessor(obj)
+	if err != nil {
+		return err
+	}
+	if !containsFinalizer(accessor.GetFinalizers(), finalizerName) {
+		return nil // resource has no finalizer, probably migration.
+	}
+	c, err := rest.RESTClientFor(&rest.Config{})
+	if err != nil {
+		return err
+	}
+
+	patch := patchSpec{
+		Op:    "replace",
+		Value: removeFinalizer(accessor.GetFinalizers(), finalizerName),
+		Path:  "/metadata/finalizers",
+	}
+	p, err := json.Marshal(patch)
+	if err != nil {
+		return err
+	}
+	res := c.Patch(types.JSONPatchType).
+		AbsPath(accessor.GetSelfLink()).
+		Body(p).
+		Do()
+	if res.Error() != nil {
+		return res.Error()
+	}
+	return nil
+}
+
+func containsFinalizer(finalizers []string, finalizer string) bool {
+	for _, f := range finalizers {
+		if f == finalizer {
+			return true
+		}
+	}
+	return false
+}
+
+func removeFinalizer(finalizers []string, finalizer string) []string {
+	for i, f := range finalizers {
+		if f == finalizer {
+			finalizers = append(finalizers[:i], finalizers[i+1:]...)
```

Comment:
This is a typical case for introducing garbage collection issues and causing processes to OOM. Please see https://github.com/golang/go/wiki/SliceTricks and lets use the safer variant of deletion.

Reply:
I don't think that is a problem here, these issues happen with pointers or structs with pointer fields. So removing a string from a string slice is fine like this.

Reply:
Ok cool. In my head I thought about the runtime objects from the k8s API, but right. These here are just simple strings.